cloud/aws/providerconfig/types.go

```diff
@@ -14,25 +14,132 @@
+	PublicIP *bool
+
+	// SecurityGroups is an array of references to security groups that should be applied to the
+	// instance.
+	SecurityGroups []AWSResourceReference
+
+	// Subnet is a reference to the subnet to use for this instance
+	Subnet AWSResourceReference
+}
+
+// AWSResourceReference is a reference to a specific AWS resource by ID, ARN, or filters.
+// Only one of ID, ARN or Filters may be specified. Specifying more than one will result in
+// a validation error.
+type AWSResourceReference struct {
+	// ID of resource
+	ID *string
+
+	// ARN of resource
+	ARN *string
+
+	// Filters is a set of filters used to identify a resource
+	Filters []Filter
+}
+
+// Filter is a filter used to identify an AWS resource
+type Filter struct {
+	// Name of the filter. Filter names are case-sensitive.
+	Name string
+
+	// Values includes one or more filter values. Filter values are case-sensitive.
+	Values []string
+}
+
+// TagSpecification is the name/value pair for a tag
+type TagSpecification struct {
+	// Name of the tag
+	Name string
+
+	// Value of the tag
+	Value string
 }
 
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 type AWSClusterProviderConfig struct {
-	metav1.TypeMeta `json:",inline"`
+	metav1.TypeMeta
 }
 
+// AWSMachineProviderStatus is the type that will be embedded in a Machine.Status.ProviderStatus field.
+// It containsk AWS-specific status information.
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 type AWSMachineProviderStatus struct {
-	metav1.TypeMeta `json:",inline"`
+	metav1.TypeMeta
+
+	// InstanceID is the instance ID of the machine created in AWS
+	InstanceID *string
+
+	// InstanceState is the state of the AWS instance for this machine
+	InstanceState *string
+
+	// Conditions is a set of conditions associated with the Machine to indicate
+	// errors or other status
+	Conditions []AWSMachineProviderCondition
+}
+
+// AWSMachineProviderConditionType is a valid value for AWSMachineProviderCondition.Type
+type AWSMachineProviderConditionType string
+
+// Valid conditions for an AWS machine instance
+const (
+	// MachineCreated indicates whether the machine has been created or not. If not,
+	// it should include a reason and message for the failure.
+	MachineCreated AWSMachineProviderConditionType = "MachineCreated"
+)
+
+// AWSMachineProviderCondition is a condition in a AWSMachineProviderStatus
+type AWSMachineProviderCondition struct {
+	// Type is the type of the condition.
+	Type AWSMachineProviderConditionType
+	// Status is the status of the condition.
+	Status corev1.ConditionStatus
+	// LastProbeTime is the last time we probed the condition.
+	// +optional
+	LastProbeTime metav1.Time
+	// LastTransitionTime is the last time the condition transitioned from one status to another.
+	// +optional
+	LastTransitionTime metav1.Time
+	// Reason is a unique, one-word, CamelCase reason for the condition's last transition.
+	// +optional
+	Reason string
```

Comment:
Why wouldn't we bubble these up to the MachineStatus ErrorReason/ErrorMessage instead of having a separate Reason/Message here?

Reply:
Conditions are meant to provide more detailed status, giving you time, condition, etc. and allowing the possibility of reporting multiple conditions at a time. If we want to provide a more concise status at the higher level, then I can add an ErrorReason and ErrorMessage there.

Reply:
There is already an ErrorReason/ErrorMessage on the MachineStatus object: https://github.com/kubernetes-sigs/cluster-api/blob/051f338bdacb76117d73a86258bc58c946add7b5/pkg/apis/cluster/v1alpha1/machine_types.go#L112-L133

It sounds like there could be a use case for duplicating the Reason/Message here as well for more granularity, though.